doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,258 @@
+# Universal, synchronous and in-thread loader hooks
+
+For background and motivation, see https://github.com/nodejs/node/issues/52219. Prototype implementation is in https://github.com/joyeecheung/node/tree/sync-hooks.
+
+TL;DR: the top priority of this proposal is to allow sun-setting CJS loader monkey patching as non-breakingly as possible (which is why it needs to be synchronous and in-thread because that's how the CJS loader and `require()` works). Then it's API consistency with the existing `module.register()` off-thread hooks.
```

Comment:
nit: I would list "in-thread" first, because we only need it to be synchronous because off-threading breaks some usecase IIUC.
```suggestion
TL;DR: the top priority of this proposal is to allow sun-setting CJS loader monkey patching as non-breakingly as possible (which is why it needs to be in-thread and synchronous because that's how the CJS loader and `require()` works). Then it's API consistency with the existing `module.register()` off-thread hooks.
```

Reply:
The sentence literally starts with:

> the top priority of this proposal is to allow sun-setting CJS loader monkey patching as non-breakingly as possible

I think both are equally important when we talk about sun-setting the CJS loader monkey patching while avoiding breakages. Off-threading is already off the table because migrating to it is bound to be a semver-major change for intermediate packages like `require-in-the-middle` or `pirates`.

Reply:
Sure, but by off-threading, we can have an async operation that appears sync to the main thread. However, that's off the table _because_ the packages you cited also expect the hooks to be in-thread.